src/ngraph/runtime/gpu/gpu_cuda_kernel_builder.cpp

```diff
@@ -181,6 +181,86 @@ void runtime::gpu::CudaKernelBuilder::get_ew_collective_op(
     return;
 }
 
+//each thread calculate the whole reduction of one output
+void runtime::gpu::CudaKernelBuilder::get_reduce_op(codegen::CodeWriter& writer,
+                                                    const std::string& name,
+                                                    runtime::gpu::GPUKernelArgs& args,
+                                                    const std::vector<std::string>& data_types,
+                                                    const std::string& reduce_op,
+                                                    size_t out_rank,
+                                                    size_t reduce_rank)
+{
+    writer << "extern \"C\" __global__ void cuda_" << name << args.get_input_signature();
+    writer.block_begin();
+    {
+        writer << "uint32_t tid = blockIdx.x * blockDim.x + threadIdx.x; \n";
+        writer << "if (tid < nthreads)\n";
+        writer.block_begin();
+        {
+            if (out_rank > 0)
+            {
+                writer << "uint32_t out_idx = tid;\n";
+            }
+            writer << "uint32_t in_idx = 0;\n";
+            writer << data_types[1] << " r = 0;\n";
+
+            //loop through all reduction axis
+            int64_t i = 0;
+            for (; i < static_cast<int64_t>(out_rank); i++)
+            {
+                writer << "in_idx += (out_idx / out_strides" << i << ") * non_reduce_in_strides"
+                       << i << ";\n";
+                writer << "out_idx %= out_strides" << i << ";\n";
+            }
+            int64_t rr = static_cast<int64_t>(reduce_rank);
+            for (int64_t j = 0; j < rr - 1; j++)
+            {
+                writer << "for(int idx" << j << " = 0; idx" << j << "< reduce_shape" << j << "; idx"
+                       << j << "++)\n";
+                writer.block_begin();
+            }
+            {
+                writer << "uint32_t reduce_idx = in_idx;\n";
+                for (int64_t j = 0; j < rr - 1; j++)
+                {
+                    writer << "reduce_idx += idx" << j << " * reduce_strides" << j << ";\n";
+                }
+                writer << "int idx" << rr - 1 << " = 0;\n";
+                writer << "uint32_t step = reduce_strides" << rr - 1 << ";\n";
+                //unroll last reduction axis
+                writer << "for(; idx" << rr - 1 << "< (reduce_shape" << rr - 1 << " >> 3); idx"
+                       << rr - 1 << "++)\n";
+                writer.block_begin();
+                {
+                    for (int k = 0; k < 8; k++)
```

Comment:
Since you always reduce on the last reduction axis, you could make the length of that axis part of the kernel and unroll it completely to avoid the remainder loop. Would help most when the length is less than 8.

Reply:
Theoretically it will help. But I actually  tested it, it make no performance difference. and the way to completely unroll will create a lot of kernels instead of only one here.

Reply:
In that case, have you tried removing the unroll completely? Wondering if the reason you didn't see any difference is that it is spending all the time waiting on those strided loads anyway.